// @rx-angular/cdk/fesm2022/cdk-ssr.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, PLATFORM_ID, DOCUMENT, NgZone, signal, Injectable } from '@angular/core';
import { toObservable } from '@angular/core/rxjs-interop';
import { isPlatformServer, isPlatformBrowser } from '@angular/common';

const defaultConfig = {
    timeout: 10000,
    logging: false,
};
/**
 * Injection token for the `HydrationTracker` service configuration.
 */
const HYDRATION_TRACKER_CONFIG_TOKEN = new InjectionToken('HYDRATION_TRACKER_CONFIG_TOKEN', {
    providedIn: 'root',
    factory: () => defaultConfig,
});
/**
 * Provides a configuration object for the `HydrationTracker` service.
 *
 * @param config - The configuration object.
 *
 * Example usage:
 * ```ts
 * import { provideHydrationTracker } from '@rx-angular/cdk/ssr';
 *
 * const appConfig: ApplicationConfig = {
 *   providers: [provideHydrationTracker({ timeout: 10000 })],
 * };
 * ```
 */
function provideHydrationTracker(config = defaultConfig) {
    return {
        provide: HYDRATION_TRACKER_CONFIG_TOKEN,
        useValue: { ...defaultConfig, ...config },
    };
}

/**
 * An injection token that provides information about the current platform.
 *
 * It provides the following information:
 * - `isServer`: Whether the current platform is the server.
 * - `isBrowser`: Whether the current platform is the browser.
 * - `isServerRenderer`: Whether the current platform is the browser and the application was server-side rendered.
 */
const PLATFORM = new InjectionToken('PLATFORM', {
    providedIn: 'platform',
    factory: () => {
        const platformId = inject(PLATFORM_ID);
        const document = inject(DOCUMENT);
        const isServer = isPlatformServer(platformId);
        const isBrowser = isPlatformBrowser(platformId);
        const isServerRendered = isBrowser && !!document.getElementById('ng-state');
        return { isServer, isBrowser, isServerRendered };
    },
});

/**
 * A high-performance utility to track application hydration status using a MutationObserver,
 * with a timeout safety net.
 *
 * It provides a signal `isFullyHydrated` that becomes true once all components are
 * hydrated or after a configurable timeout (default: 10 seconds).
 *
 * Disclaimer: The service only runs in the browser and is not available on the server.
 */
class HydrationTracker {
    constructor() {
        this.config = inject(HYDRATION_TRACKER_CONFIG_TOKEN);
        this.platform = inject(PLATFORM);
        this.ngZone = inject(NgZone);
        this.observer = null;
        this.timeoutId = null; // Stores the setTimeout ID
        /**
         * A reactive signal that emits `true` when the application is fully hydrated.
         */
        this.isFullyHydrated = signal(false, ...(ngDevMode ? [{ debugName: "isFullyHydrated" }] : []));
        /**
         * An observable that emits `true` when the application is fully hydrated.
         */
        this.isFullyHydrated$ = toObservable(this.isFullyHydrated);
        if (this.platform.isServerRendered) {
            this.initializeObserver();
        }
        else if (this.platform.isBrowser) {
            this.isFullyHydrated.set(true);
        }
    }
    initializeObserver() {
        const unhydratedElements = this.getUnhydratedElements();
        let unhydratedCount = unhydratedElements.length;
        if (unhydratedCount === 0) {
            this.isFullyHydrated.set(true);
            if (this.config.logging) {
                console.log('[HydrationTracker] ✅ Application was already hydrated on initialization.');
            }
            return;
        }
        this.observer = new MutationObserver((mutations) => {
            for (const mutation of mutations) {
                if (mutation.type === 'attributes' &&
                    mutation.attributeName === 'ngh') {
                    const element = mutation.target;
                    if (!element.hasAttribute('ngh')) {
                        unhydratedCount--;
                    }
                }
            }
            if (unhydratedCount <= 0) {
                // Hydration finished normally, so we don't need the timeout.
                this.completeHydration(false);
            }
        });
        unhydratedElements.forEach((element) => {
            this.ngZone.runOutsideAngular(() => {
                this.observer?.observe(element, {
                    attributes: true,
                    attributeFilter: ['ngh'],
                });
            });
        });
        if (this.config.timeout) {
            // Set a timeout as a safety net.
            this.timeoutId = this.ngZone.runOutsideAngular(() => setTimeout(() => {
                this.completeHydration(true); // `true` indicates it was a timeout
            }, this.config.timeout));
        }
    }
    getUnhydratedElements() {
        return Array.from(document.body.querySelectorAll('[ngh]'));
    }
    completeHydration(timedOut) {
        // If this function is running, we must clear any pending timeout.
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
        this.isFullyHydrated.set(true);
        if (timedOut) {
            if (this.config.logging) {
                console.warn(`[HydrationTracker] 🟡 Hydration check timed out after ${this.config.timeout} milliseconds. Forcing completion.`);
            }
        }
        else {
            if (this.config.logging) {
                console.log('[HydrationTracker] ✅ Application is now fully hydrated.');
            }
        }
        // Disconnect the observer to free up resources.
        this.observer?.disconnect();
        this.observer = null;
    }
    ngOnDestroy() {
        // Ensure both the observer and the timeout are cleaned up.
        this.observer?.disconnect();
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
        }
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.0.6", ngImport: i0, type: HydrationTracker, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "21.0.6", ngImport: i0, type: HydrationTracker, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.0.6", ngImport: i0, type: HydrationTracker, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { HYDRATION_TRACKER_CONFIG_TOKEN, HydrationTracker, PLATFORM, provideHydrationTracker };
